Add Y_lm() function to hydrogen.py

weisscharlesj opened the pull request:
Adds the spherical harmonic function Y_lm() to sympy.physics.hydrogen.py. 

### What's Done:
- Added the spherical harmonic function Y_lm() to sympy.physics.hydrogen.py to provides easy user access to angular wave functions with a signature consistent with Psi_nlm() and R_nl() in physics.hydrogen.py and adds documentation and examples.

### Code Added:

`from sympy.functions.special.spherical_harmonics import Ynm, Ynm_c`

```
def Y_lm(l, m, phi, theta, complex=False):
    """
    Returns the spherical harmonic or angular function Y_{l}^{m} of atomic
    orbitals.

    Parameters
    ==========

    l : integer
        ``l`` is the Angular Momentum Quantum Number with
        values ranging from 0 to ``n-1``.
    m : integer
        ``m`` is the Magnetic Quantum Number with values
        ranging from ``-l`` to ``l``.
    phi :
        azimuthal angle
    theta :
        polar angle
    complex : bool
        whether the function is complex or not

    Examples
    ========

    >>> import sympy
    >>> from sympy.physics.hydrogen import Y_lm
    >>> phi, theta = sympy.symbols('phi theta')
    >>> Y_lm(1, 0, phi, theta)
    sqrt(3)*cos(theta)/(2*sqrt(pi))
    >>> Y_lm(2, 0, phi, theta).simplify()
    sqrt(5)*(3*cos(theta)**2 - 1)/(4*sqrt(pi))
    >>> Y_lm(2, 1, phi, theta, complex=False)
    -sqrt(30)*exp(I*phi)*sin(theta)*cos(theta)/(4*sqrt(pi))
    >>> Y_lm(2, 1, phi, theta, complex=True)
    -sqrt(30)*exp(-I*phi)*sin(theta)*cos(theta)/(4*sqrt(pi))
    >>> Y_lm(1, 1, 1, 1).evalf()
    -0.157078470548806 - 0.244635223409689*I

    Find angular nodes in dz2 atomic orbital by solving for zero probability.

    >>> dz2 = Y_lm(2, 0, phi, theta)**2
    >>> [sol.evalf() for sol in sympy.solve(dz2)]
    [4.0969092717143, 5.32786868905508, 2.18627603546528, 0.955316618124509]

    """
    l, m, theta, phi = map(S, [l, m, theta, phi])
    if abs(m) > l:
        raise ValueError("|'m'| must be less or equal 'l'")
    elif m.is_integer and l.is_integer:
        if complex:
            return Ynm_c(l, m, theta, phi).expand(func=True)
        else:
            return Ynm(l, m, theta, phi).expand(func=True)
    else:
        raise ValueError("'m' and 'l' must be integers")
```
Release Notes
<!-- BEGIN RELEASE NOTES -->
* physics.hydrogen
    * Added the spherical harmonic function Y_lm() to sympy.physics.hydrogen.py
<!-- END RELEASE NOTES -->

weisscharlesj committed:
Add Y_lm() function to hydrogen.py

Add spherical harmonic function Y_lm() to sympy.physics.hydrogen.py. This provides easy access to angular wave functions with a signature consistent with Psi_nlm and R_nl. Documentation and examples are also added.

weisscharlesj committed:
Fix import error

Add 'import sympy' to example to fixe import error.

weisscharlesj committed:
Fix import Y_lm error

Add 'from sympy.physics.hydrogen import Y_lm' to docstring.

weisscharlesj committed:
Remove 0 from docstring test

Remove trailing zero from docstring example to fix failed doctest.

weisscharlesj committed:
Update .mailmap file

weisscharlesj committed:
Update .mailmap file

oscarbenjamin commented on sympy/physics/hydrogen.py:
```diff
@@ -92,6 +92,63 @@ def R_nl(n, l, r, Z=1):
+    """
+    l, m, theta, phi = map(S, [l, m, theta, phi])
+    if abs(m) > l:
+        raise ValueError("|'m'| must be less or equal 'l'")
```
This condition will fail if `m` or `l` is a symbol.

oscarbenjamin replied:
I don't know if that is just expected here as I have not used this module.

weisscharlesj replied:
The `abs(m) > l` check now only occurs if both `l` and `m` are integers.
```
if m.is_integer and l.is_integer and abs(m) > l:
        raise ValueError("|'m'| must be less or equal 'l'")
```

weisscharlesj committed:
Allow Y_lm to support non-integers for l and m

weisscharlesj committed:
Add tests and support for real representations

Add tests to the test_hydrogen.py file and support for real representations by calling Znm() from the spherical harmonics submodule.

oscarbenjamin commented on sympy/physics/hydrogen.py:
```diff
@@ -92,6 +92,69 @@ def R_nl(n, l, r, Z=1):
+
+    """
+    l, m, theta, phi = map(S, [l, m, theta, phi])
+    if m.is_integer and l.is_integer and abs(m) > l:
+        raise ValueError("|'m'| must be less or equal 'l'")
```
This can still fail for symbols:
```python
In [7]: x = symbols('x', integer=True)

In [8]: x.is_integer
Out[8]: True

In [9]: if x > 1:
   ...:     print('greater')
   ...: 
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
...
TypeError: cannot determine truth value of Relational: x > 1
```
The usual way to write this would be:
```
if (abs(m) > l) is S.true:
```
That one can still fail for complex `l` but I don't know if that is relevant here.

weisscharlesj replied:
Okay, I see what you mean. I will remove the check to allow the `Ynm` and `Znm` functions to perform any checks per discussion below.

oscarbenjamin commented on sympy/physics/hydrogen.py:
```diff
@@ -92,6 +92,69 @@ def R_nl(n, l, r, Z=1):
+    l, m, theta, phi = map(S, [l, m, theta, phi])
+    if m.is_integer and l.is_integer and abs(m) > l:
+        raise ValueError("|'m'| must be less or equal 'l'")
+    if real:
+        return Znm(l, m, theta, phi).expand(func=True)
+    else:
+        return Ynm(l, m, theta, phi).expand(func=True)
```
I don't know these functions so well but is it useful to have a function `Y_lm` that is just an alias for `Znm` and `Ynm` like this?

Usually I would say that if you have a boolean parameter like `Y_lm(..., real=True/False)` then it is better just to have two separate functions but then we would be back to these exact two functions.

Is it better here just to document these functions somehow rather than adding a new function that is an alias for existing functions?

weisscharlesj replied:
Sounds good. I will remove the checks to allow `Ynm` and `Znm` to perform any checks and split `Y_lm` into two functions: `Y_lm` for complex and `Z_lm` for real. I believe this should resolve all noted issue. Thank you for your input!

weisscharlesj committed:
Separate  Y_lm into two functions

Separate Y_lm into two functions: Y_lm and Z_lm for real. Also removed checks and sympification allowing Ynm and Znm to handle this instead. Updated examples and tests accordingly.

weisscharlesj committed:
Add See Also and References sections

Add See Also sections to Y_lm and Z_lm along with references to these functions from Ynm, Ynm_c, and Znm. Also added references to Y_lm and Z_lm functions.

weisscharlesj commented on sympy/physics/hydrogen.py:
```diff
@@ -92,6 +92,109 @@ def R_nl(n, l, r, Z=1):
+
+    This function follows the Condon-Shortley phase convention and may return
+    real or complex values depending on values for ``l`` and ``m``. See Z_lm()
+    for real-valued spherical harmonics.
```
I added See Also sections to both Y_lm and Z_lm functions and references to these function from the Ynm, Ynm_c, and Znm functions as suggested. I also added references while I was rounding out the docstrings and verified that older reference links from Ynm were still good.
